spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/hazelcast/HazelcastServerConfiguration.java

```diff
@@ -93,14 +98,43 @@ private static Config loadConfig(URL configUrl) throws IOException {
 	@Configuration(proxyBeanMethods = false)
 	@ConditionalOnSingleCandidate(Config.class)
 	static class HazelcastServerConfigConfiguration {
-
 		@Bean
 		HazelcastInstance hazelcastInstance(Config config) {
 			return getHazelcastInstance(config);
 		}
 
 	}
 
+	private interface ConfigurationCustomizer {
+		ConfigurationCustomizer EMPTY_CUSTOMIZER = c -> c;
+		Config customize(Config configuration);
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	static class HazelcastConfigCustomizerConfiguration {
+		private final ApplicationContext applicationContext;
+
+		public HazelcastConfigCustomizerConfiguration(ApplicationContext applicationContext) {
+			this.applicationContext = applicationContext;
+		}
+
+		@Bean
+		@ConditionalOnClass(name = "com.hazelcast.spring.context.SpringManagedContext")
+		public ConfigurationCustomizer springManagedContextConfigurationCustomizer() {
+			return configuration -> {
+				SpringManagedContext springManagedContext = new SpringManagedContext(applicationContext);
+				configuration.setManagedContext(springManagedContext);
+				return configuration;
+			};
+		}
+
+		@Bean
+		@ConditionalOnMissingBean(ConfigurationCustomizer.class)
```

Comment:
I'm not sure if this is the right way to introduce a fallback. 

It appears it's dependant on the ordering of methods in a class. If I move this method above `springManagedContextConfigurationCustomizer()` then both beans are constructed. So I reckon there is a less fragile way, any hint is appreciated.

Reply:
I guess the trick is to use a separate configuration class, take a look: https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jdbc/DataSourceAutoConfiguration.java#L66-L72

Reply:
Thanks for this suggestion, I was considering it, but it looked like an overkill to me. Too much ceremony, for my taste anyway. 

Given the other changes `EMPTY_CUSTOMIZER` is not needed anymore the ordering of bean factory methods is not a concern anymore.